packages/rancher-sachet/charts/files/pre-install-job.sh

```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -e
+
+namespace="${NAMESPACE}"
+fullname="${FULLNAME}"
+preName="${PRENAME}"
+
+# exit if the configmap already exists
+kubectl get configmap -n ${namespace} ${fullname} > /dev/null 2>&1 && { echo "configMap already exists"; exit  1; }
+
+kubectl patch configmap -n ${namespace} ${preName} --dry-run=client -o yaml -p "$(printf "{\"metadata\":{\"name\": \"%s\"}}" ${fullname} )" | kubectl apply -f -
+kubectl annotate configmap -n ${namespace} ${fullname} helm.sh/hook- helm.sh/hook-delete-policy- helm.sh/hook-weight-;
```

Comment:
I'm not convinced this whole job is actually needed. Can you explain what exactly the patch is doing, its grabbing `configmap-pre-install.yaml` and then changing name to `{fullname}` then installing it. Why not just use the proper name of the configmap and then have helm error out if it already exists ? That logic also exists at the top of the yaml itself. Then I think you could drop the annotate too. Was there a reason you went this way ?

Reply:
By using this patch, we make the configmap independent from the app.
The pre-install hook will create the default configmap if it is the first time to install the app and the configmap does not exist in the target namespace. Since K8s does not allow us to change the name of a resource, so we need to create a new configmap by patching the existing one with a new name. 
Finally, helm deletes the configmap created by the pre-install hook, so we have only the new configmap left.


If it is a `helm install/upgrade/delete` operation, the configmap will NOT be touched since it is not part of the app. 

^^^ @aiyengar2  Please correct me if the description is not accurate.


@cbron  @aiyengar2  I think we can use helm's annotation `"helm.sh/resource-policy": keep` to achieve the same result. 
1. add the annotation `"helm.sh/resource-policy": keep` so the config will be kept
2. add a check that skip creating the configmap if it already exists

Reply:
@jiaqiluo perfect explanation. In a TLDR, what we seek to accomplish is Helm only installing a resource, but never upgrading or deleting it. 

Helm only currently supports disabling the deletion of the resource via a Helm annotation (e.g. `"helm.sh/resource-policy": keep`), but does not support disabling upgrades by definition, which is why I think the annotation based approach may not work.

Also, re: `adding a check that skips creating the ConfigMap if it already exists`, I think Helm will identify that as a need to delete the resource on the upgrade (since the resource is no longer tracked by the deployed manifest), but if you can get that to work that would be ideal and something we should also introduce in the main monitoring chart to avoid having to use this image at all.

If that doesn't work, here's another thing that we could try that is way easier:
- Place each ConfigMap under the `files/default-config-maps` directory (so it does not get treated as a template)
- Create a ConfigMap `preinstall-default-configmaps` whose keys are all of the files in `files/default-config-maps` like we do for [dashboards](https://github.com/rancher/charts/blob/80b8a038fbd99e104f7862d31f8b4c629450c1a7/charts/rancher-monitoring/rancher-monitoring/14.5.100-rc05/templates/rancher-monitoring/dashboards/rancher/linux-dashboards.yaml#L16), but apply `tpl` on the contents emitted to ensure we can still use Helm's templating engine to fill in the values in the files.
- Mount the `preinstall-default-configmaps` ConfigMap onto the pre-install / pre-upgrade job at `/some/path`. Within this Job, run through a for loop that runs `kubectl create /some/path/<config-map>.yaml` on each of the ConfigMaps, ignoring (but logging) errors that are encountered when running `kubectl create`. 
 
Since `kubectl create` will fail if the resource already exists, this may be a better way of achieving what we want from this.

Reply:
>  In a TLDR, what we seek to accomplish is Helm only installing a resource, but never upgrading or deleting it.

@aiyengar2 I can achieve this goal by the combination of the annotations 
```
"helm.sh/hook": pre-install 
"helm.sh/resource-policy": keep
```
with the checks for existence:
```
{{- if not (lookup "v1" "ConfigMap" (include "sachet.namespace" . ) (include "sachet.fullname" .)) }}
// create the configmap
{{- end }}
```

Reply:
Yes using both of those annotations should get what we want. I'm not positive but I think pre-install does not get touched after an install, otherwise there would be no reason for the `pre-upgrade` hook, same for delete I suppose. But @jiaqiluo if you haven't already we should definitely test those scenarios.

I want to confirm the above, it seems all we need is: 
a) if the configmap pre-exists we don't do anything
b) we install configmap on create
c) we don't modify again, for example on upgrade or delete

My thinking is they are solved by: 
a) helm ownership policy. We may also need conditional in file though if we do install, uninstall, install. Need to test.
b) pre-install hook
c) no other hooks

Reply:
Actually now that I consider the install -> uninstall -> install flow, I'm not sure part A will work. We do want to be able to re-install and use existing file. Need to test that flow and see if helm errors out completely.

Reply:
I have validated that the current solution supports all the scenarios you mentioned :)